### ISSUE TITLE
Support learn() with total timesteps less than episode length

### ISSUE THREAD

**Issue by shwang** [hill-a/stable-baselines#619]
In an imitation learning project, I'm alternating short calls (low `total_timesteps`) to `PPO2.learn()` with gradient descent updates to the discriminator among other things.

The PPO2 updates themselves seem to be working fine. However, logging has a few problems because some logging state is not kept between different calls to `learn()`. Despite the fact that I use `.learn(reset_num_timesteps=False)`, which seems to exist to allow logging between multiple calls to `PPO2.learn()`, we run these problems:

(1) We initialize a new Runner every time we call `learn`, even if `reset_num_timesteps=False`. This forces the environment to reset(), thus biasing the logger towards reporting the reward and episode length means of shorter episodes (longer episodes don't get to finish by the end of the training loop). As an example, when training CartPole, my expert imitation policies (mean return: 500) often show mean return around 300 in my PPO2 logs.

(2) The `ep_info_buf` is reset every time we call learn instead of being saved. This makes the training curve more jagged and again biased toward shorter episodes at the beginning of each new call to `learn()`.

I'm wondering if the maintainers would be interested in some sort of PR that fixes these logging discrepancies.

I propose that `reset_num_timesteps=False` (perhaps rename to `reset_log_state`) should make PPO2 keep state from (1) and (2).

**Comment by araffin** [comment]
>I'm wondering if the maintainers would be interested in some sort of PR that fixes these logging discrepancies.

Hello,
Sounds good to improve the consistency, but I would do that for the other that use a runner too (A2C, ACER, TRPO and PPO1).

>perhaps rename to reset_log_state

I'm not sure about that (this would mean different API depending on the algorithm)

**Comment by AdamGleave** [comment]
> Sounds good to improve the consistency, but I would do that for the other that use a runner too (A2C, ACER, TRPO and PPO1).

I think similar problems exist in other algorithms too -- e.g. DQN has internal `episode_rewards` that is an instance variable of `learn()`.

Ideally we'd fix this everywhere, but it seems OK to me to do this piecemeal. All the algorithms using Runner's are a natural starting point as should be similar code changes.

> 
> > perhaps rename to reset_log_state
> 
> I'm not sure about that (this would mean different API depending on the algorithm)

Agree with @araffin we should keep consistent interface between algorithms. I'd leave it as `reset_num_timesteps` for now. If in the future we fix all the algorithms then can maybe rename it.

**Comment by shwang** [comment]
I'm happy to go through all the algorithms that use runner and make sure they don't make a new Runner with each call to learn.

Not sure if I will have enough understanding from skimming to make sure that all the pertinent log state is maintained between calls to `<model>.learn()` though, so I'm only preparing to change the episode reward buffer on PPO2.

**Comment by AdamGleave** [comment]
After talking to @shwang the problem is more serious than I thought: it affects learning not just logging. `AbstractEnvRunner` calls `env.reset` in the `__init__` method. So any algorithm using `Runner` when you call `learn` with fewer timesteps than the episode length will never see the episode end during training! This would cause big problems in environments with sparse reward at the end of an episode.

Admittedly this use case is rare so it's fairly low-severity, but think we should treat learning breaking in this setting as a bug.

**Comment by shwang** [comment]
edge case: Runner.__init__ takes in the environment, so it's invalidated whenever we call `set_env`. I'm going to force a Runner reset whenever `set_env` is called.

**Comment by shwang** [comment]
Hi @AdamGleave and @araffin, wanted to check-in on some design decisions and have a question about the segment generator utility in TRPO.

My branch so far can be found at #639. 

To make sure that Runner is only initialized when:
(1) The ActorCriticRLModel is initialized OR
(2) The environment is changed via `set_env` (thereby invalidating `self.runner`)

I've moved the runner initialization logic into an abstract method `ActorCriticRLModel._make_runner` and call this function from within `ActorCriticRLModel.__init__` and `ActorCriticRLModel.set_env`:

https://github.com/hill-a/stable-baselines/blob/8d7a364226383e7c1383ae0e7c16fbed87fa748d/stable_baselines/common/base_class.py#L893-L915

(The PR currently implements this logic in a temporary class `ActorCriticWithRunnerRLModel` because `PPO1` and `TRPO` aren't compatible with this interface right now. More on that later/)

Does this seem okay? Every `ActorCriticRLModel` uses a Runner (or a pseudo-Runner in the case of TRPO, PPO1) of some sort right now.

**Comment by shwang** [comment]
My plan on `traj_segment_generator` used in TRPO and PPO1:

I would like to convert this generator function into a `AbstractEnvRunner` in order to allow the aforementioned algorithms to fit the `ActorCriticRlModelWithRunner` interface so I don't have to have separate classes `ActorCriticRLModel{,WithRunner}`.

Probably would call it `TRPORunner`. This would be different from all the other Runner in that it would have a different return signature for `run()`. `traj_segment_generator` yields a huge dictionary. All the current `Runner`s return something like
`                obs, returns, masks, actions, values, neglogpacs, states, ep_infos, true_reward = self.runner.run()
`.

Would rather modify `TRPORunner`'s logic as little as possible so as not to accidentally introduce errors.

Does changing `traj_segment_generator` into a Runner seem fine?

**Comment by AdamGleave** [comment]
> Does this seem okay? Every `ActorCriticRLModel` uses a Runner (or a pseudo-Runner in the case of TRPO, PPO1) of some sort right now.

This seems reasonable to me. It's worth thinking about whether `ActorCriticRLModel` inherently has a `Runner` (i.e. any future algorithms implemented in this class will too), or if it's an artifact of the current implementation. It does seem to me like any kind of policy gradient or AC algorithm will need a runner, but @araffin may have additional thoughts.

Out of interest, would anything bad happen if `_make_runner` were just defined as a no-op in the concrete class?

> Does changing traj_segment_generator into a Runner seem fine?

This does seem quite messy but the whole TRPO and PPO1 codebases aren't great, so there's unlikely to be a clean solution that doesn't involve a lot of refactoring. I'd be in favour unless @araffin has a cleaner idea.

**Comment by shwang** [comment]
> Out of interest, would anything bad happen if _make_runner were just defined as a no-op in the concrete class?

No, as of now this would just set `self.runner = None`.

**Comment by araffin** [comment]
@AdamGleave @shwang I will try to take a look at that today or next week. Quick remark: I plan to remove the runners in the V3 anyway (and replace them with a method `collect_rollouts` (one for on-policy/ one for off-policy) so we can keep track of all the variables inside the class)

**Comment by shwang** [comment]
@araffin Depending on how long you expect it to take to remove the runners and replace them with `collect_rollouts`, it might make sense for me not to refactor TRPO's `utils.traj_segment_generator` into a `Runner`.

But maybe this refactoring might be helpful for your later change, and it might be awkward to add in a new base class `ActorCriticRLModelWithRunner` even if temporarily.

**Comment by shwang** [comment]
Refactoring TRPO's `traj_segment_generator` into a Runner correctly seems like a huge pain. I'm halfway through right now.

Would really appreciate if could give me some guidance on how some steps I can take to confirm that I haven't made any unintended changes to TRPO. Unit tests probably aren't sufficient.

**Comment by araffin** [comment]
@shwang sorry, I was more busy than expected... also this issue seems more complex than it appears.
I agree that having `ActorCriticWithRunnerRLModel` is not super great.

But maybe we should decouple the PR into parts:
1. A PR that solves the issue for A2C/PPO2/ACKTR/ACER
2. A PR that solves the issue for PPO1/TRPO
3. other algorithms?

so we can focus on one thing and handle more complex things (PPO1/TRPO refactoring) afterward.

>Depending on how long you expect it to take to remove the runners and replace them with collect_rollouts

I don't know yet. I have other priority for now (writing the callback collection and starting a first tf2 version), so don't expect that to be done soon.

**Comment by shwang** [comment]
@araffin 

> sorry, I was more busy than expected
No prob

> But maybe we should decouple the PR into parts
Seems good to me, #639 should fix the issue for A2C/PPO2/ACKTR/ACER. I've also put in the comments of the test I made to fix this bug some TODOs for part 2 and part 3.

**Comment by Solliet** [comment]
This issue exists in SAC as well. In particular, the rewards are stored in a list that is reset every time `.learn` is called.

**Comment by araffin** [comment]
This should be fixed in v3, but needs to be checked, linking https://github.com/DLR-RM/stable-baselines3/issues/1
